src/main/java/io/gingersnapproject/cdc/cache/hotrod/HotRodService.java

```diff
@@ -44,64 +43,60 @@ public class HotRodService implements CacheService {
 
    @Inject DBSyncerMetrics metrics;
 
-   volatile RemoteCacheManager rcm;
-
-   private final Map<String, CacheBackend> backends = new ConcurrentHashMap<>();
+   private final Map<CacheIdentifier, HotRodCache> backends = new ConcurrentHashMap<>();
    private volatile boolean shutdownSignal;
 
-   @PostConstruct
-   public void init() {
-      rcm = new RemoteCacheManager(config.cache().uri());
-      createCaches();
-   }
-
    public void shutdown(@Observes ShutdownEvent ignore) {
-      if (rcm != null) {
-         shutdownSignal = true;
-         if (backends.values().stream().noneMatch(CacheBackend::isRunning)) destroy();
-      }
+      shutdownSignal = true;
+      if (backends.values().stream().noneMatch(c -> c.backend.isRunning())) destroy();
    }
 
    @Override
-   public CompletionStage<Boolean> reconnect(String name, Rule rule) {
-      if (!backends.containsKey(name)) return CompletableFutures.completedFalse();
-
-      return CompletableFuture.supplyAsync(() -> {
-         backendForRule(name, rule).start();
-         return null;
-      }).handle((ignore, t) -> {
-         if (t != null) {
-            eventing.backendFailedEvent(name, t);
-            return false;
-         }
-         createCaches();
-         return true;
-      });
+   public CompletionStage<Boolean> reconnect(CacheIdentifier identifier, Rule rule) {
+      var cache = getCacheOrThrow(identifier);
+      return cache.rcm.startAsync().thenApply(ignore -> cache.backend.reconnect());
    }
 
    @Override
-   public void stop(String name) {
-      if (!backends.containsKey(name)) throw new IllegalStateException(String.format("Backend for %s not created", name));
-      backends.computeIfPresent(name, (ignore, cache) -> {
-         cache.stop();
-         return cache;
-      });
-
-      if (shutdownSignal) destroy();
+   public void stop(CacheIdentifier identifier) {
+      var cache = getCacheOrThrow(identifier);
+      cache.backend.stop();
+      destroy();
    }
 
    private void destroy() {
-      CompletionStage<Void> cs = rcm.stopAsync();
-      rcm = null;
+      if (!shutdownSignal || backends.values().stream().anyMatch(c -> c.backend.isRunning()))
+         return;
+
+      CompletionStage<Void> cs = CompletableFutures.completedNull();
+      for (HotRodCache cache : backends.values()) {
+         cs = cs.thenCompose(ignore -> cache.rcm.stopAsync());
+      }
+      backends.clear();
       CompletableFutures.uncheckedAwait(cs.toCompletableFuture());
    }
 
    @Override
-   public CacheBackend backendForRule(String name, Rule rule) {
-      return backends.computeIfAbsent(name, ignore -> createBackend(name, rule));
+   public CacheBackend backendForRule(CacheIdentifier identifier, Rule rule) {
+      return backends.computeIfAbsent(identifier, ignore -> createBackend(identifier, rule))
+            .backend();
+   }
+
+   private HotRodCache createBackend(CacheIdentifier identifier, Rule rule) {
```

Comment:
Hrmm, I am not a fan of having multiple RemoteCacheManagers per rule if they have the same url. I thought we were just going to use different caches but reuse the same remote cache manager per url? Otherwise we are going to have a huge influx of netty threads.

TBH, thinking about this more, we probably should really look into implementing injection of the event loop for remote cache managers, so we can reuse them. This was added in ISPN-12879

Reply:
Makes sense. I'll be looking into this and applying some other changes so we can easily identify the cache manager from the stores (offset, schema, and the cache itself).